osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs

```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
                 length * MathF.Sin(angle)
             );
         }
+
+        public static bool IsHitObjectOnBeat(OsuBeatmap beatmap, OsuHitObject hitObject, bool downbeatsOnly = false)
+        {
+            var timingPoints = beatmap.ControlPointInfo.TimingPoints;
+            var currentTimingPoint = timingPoints.Reverse().FirstOrDefault(p => p.Time <= hitObject.StartTime);
+
+            if (currentTimingPoint == null)
+                return false;
+
+            double timeSinceTimingPoint = hitObject.StartTime - currentTimingPoint.Time;
+
+            double length = downbeatsOnly
+                ? currentTimingPoint.BeatLength * currentTimingPoint.TimeSignature.Numerator
+                : currentTimingPoint.BeatLength;
+
+            return (timeSinceTimingPoint + 1) % length < 2;
+        }
+
+        public static float GetRelativeTargetAngle(float targetDistance, float offset, bool flowDirection)
+        {
+            float angle = (float)(3 / (1 + 200 * Math.Pow(MathHelper.E, 0.016 * (targetDistance - 466))) + 0.45 + offset);
+            float relativeAngle = MathHelper.Pi - angle;
+            return flowDirection ? -relativeAngle : relativeAngle;
+        }
+
+        public static float RandomGaussian(Random rng, float mean = 0, float stdDev = 1)
+        {
+            double x1 = 1 - rng.NextDouble();
+            double x2 = 1 - rng.NextDouble();
```

Comment:
I get that this is the [Box-Muller transform](https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform) but why `1 -`? `rng.NextDouble()` is already distributed onto the unit interval, is it not?

Reply:
`rng.NextDouble()` returns a random number in the interval `[0;1)`, but we need `(0;1]` because `log(0)` is undefined.

Reply:
Please add a comment mentioning this.